fix: Restrict Python version to <3.13

martimlobao opened the pull request:
- fixes #1539
- enforce Python <3.13

vincentcox commented:
![image](https://github.com/user-attachments/assets/1d823ec4-5864-43a0-96a5-300421c14fc0)

martimlobao commented:
I added back the `poetry.lock` file which seems to have been accidentally deleted (at least I couldn't find any PR or commit message explaining why it's gone) and also added python 3.9 and 3.11 to the CI checks since they're both supported versions (also not sure why they're not there already, but feel free to revert the last commit if there's some reason).

vincentcox commented:
![image](https://media0.giphy.com/media/v1.Y2lkPTc5MGI3NjExaHhhN2N4ZDZpZ2U0bWI1MWE1YjY1bHh0M292Z3F3bjhpd3FxemVjNCZlcD12MV9pbnRlcm5hbF9naWZfYnlfaWQmY3Q9Zw/Q6WPVzFU8LcBWWgQE1/giphy.webp)
